File: FinalProject/app/src/main/java/com/example/finalproject/entities/Enemy.java
package com.example.finalproject.entities;

import android.graphics.Bitmap;
import android.os.Parcelable;
import android.util.Pair;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.example.finalproject.R;
import com.example.finalproject.fragments.MainActivity;
import com.example.finalproject.fragments.MapFragment;
import com.example.finalproject.fragments.MenuFragment;
import com.example.finalproject.fragments.StatusBarFragment;
import com.example.finalproject.items.Item;
import com.example.finalproject.service.Triplex;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class Enemy extends Entity implements Parcelable {
    private ArrayList<Triplex<Item, Integer, Integer>> drop;
    private int defence = 0;
    private boolean t = true;
    private Bitmap texture;
    private boolean duel, dead;

    public Enemy(String name, int health, int mana, int damage, int armor, int givenGold, int givenExp, ArrayList<Triplex<Item, Integer, Integer>> drop, Bitmap b) {
        setHealthRegen(2);
        setArmor(armor);
        setDamage(damage);
        setHealth(health);
        setName(name);
        setMaxHealth(health);
        setMaxMana(mana);
        setDrop(drop);
        setGivenExp(givenExp);
        setGivenGold(givenGold);
        setTexture(b);
    }

    public Enemy(DatabaseReference ref, Bitmap b, TextView enemyHealth, TextView enemyMana, ProgressBar p){
        ref.child("name").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setName(task.getResult().getValue(String.class));
            }
        });
        ref.child("health").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setHealth(task.getResult().getValue(Double.class));
                ref.child("maxHealth").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                        if (task.isSuccessful()&&task.getResult().getValue()!=null)
                            setMaxHealth(task.getResult().getValue(Double.class));
                        enemyHealth.setText(Math.round(getHealth())+"/"+Math.round(getMaxHealth()));
                    }
                });
            }
        });
        ref.child("mana").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setMana(task.getResult().getValue(Double.class));
                ref.child("maxMana").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                        if (task.isSuccessful()&&task.getResult().getValue()!=null)
                            setMaxMana(task.getResult().getValue(Double.class));
                        enemyMana.setText(Math.round(getMana())+"/"+Math.round(getMaxMana()));
                    }
                });
            }
        });
        ref.child("damage").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setDamage(task.getResult().getValue(Integer.class));
            }
        });
        ref.child("armor").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setArmor(task.getResult().getValue(Integer.class));
            }
        });
        ref.child("givenGold").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setGivenGold(task.getResult().getValue(Integer.class));
            }
        });
        ref.child("givenExp").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull @NotNull Task<DataSnapshot> task) {
                if (task.isSuccessful()&&task.getResult().getValue()!=null)
                    setGivenExp(task.getResult().getValue(Integer.class));
                p.setVisibility(View.GONE);
            }
        });
        ref.child("dead").addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull @NotNull DataSnapshot snapshot) {
                MainActivity.player.addExperience(getGivenExp());
                MainActivity.player.addGold(getGivenGold());
            }

            @Override
            public void onCancelled(@NonNull @NotNull DatabaseError error) {

            }
        });
        setDrop(new ArrayList<>());
        MainActivity.player.setEnemy(this);
        setTexture(b);
        duel=true;
    }

    public Enemy(Enemy enemy) {
        setHealthRegen(enemy.getHealthRegen());
        setManaRegen(enemy.getManaRegen());
        setArmor(enemy.getArmor());
        setDrop(enemy.getDrop());
        setDamage(enemy.getDamage());
        setExperience(enemy.getExperience());
        setExperienceToNextLevelRequired(enemy.getExperienceToNextLevelRequired());
        setHealth(enemy.getHealth());
        setLevel(enemy.getLevel());
        setMana(enemy.getMana());
        setMaxHealth(enemy.getMaxHealth());
        setMaxMana(enemy.getMaxMana());
        setName(enemy.getName());
        setPowerLevel(enemy.getPowerLevel());
        setGivenExp(enemy.getGivenExp());
        setGivenGold(enemy.getGivenGold());
        setTexture(enemy.getTexture());
    }

    public void attack(Player player) {
        player.take_damage(super.getDamage());
    }

    public void fight() {
        if (t) {
            attack(MainActivity.player);
            t = false;
        } else {
            defend();
            t = true;
        }
    }

    public void defend() {

    }

    public ArrayList<Triplex<Item, Integer, Integer>> getDrop() {
        return drop;
    }

    public void setDrop(ArrayList<Triplex<Item, Integer, Integer>> drop) {
        this.drop = drop;
    }

    public Bitmap getTexture() {
        return texture;
    }

    public void setTexture(Bitmap texture) {
        this.texture = texture;
    }

    public boolean isDuel() {
        return duel;
    }

    public void setDuel(boolean duel) {
        this.duel = duel;
    }

    public int getDefence() {
        return defence;
    }

    public void setDefence(int defence) {
        this.defence = defence;
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/service/classes/entities/Enemy.kt
package com.example.finalproject.service.classes.entities

import com.example.finalproject.service.classes.Damage
import com.example.finalproject.service.classes.Loot
import com.example.finalproject.service.classes.Resistances
import com.example.finalproject.service.interfaces.Dmg
import com.example.finalproject.service.interfaces.Health
import com.example.finalproject.service.serializers.EnemySerializer
import com.google.firebase.database.DatabaseReference
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.HashMap

@Serializable(with = EnemySerializer::class)
class Enemy(
    name: String,
    id: Int,
    health: Double,
    maxHealth: Double,
    healthRegen: Double,
    mana: Double,
    maxMana: Double,
    manaRegen: Double,
    resistances: Resistances,
    loot: Loot,
    override var damage: Damage
) :
    Entity(
        name = name,
        id = id,
        health = health,
        maxHealth = maxHealth,
        healthRegen = healthRegen,
        mana = mana,
        maxMana = maxMana,
        manaRegen = manaRegen,
        resistances = resistances,
        loot = loot
    ), Dmg {
    override var def: Boolean = false

    constructor(enemy: Enemy) : this(
        enemy.name,
        enemy.id,
        enemy.health,
        enemy.maxHealth,
        enemy.healthRegen,
        enemy.mana,
        enemy.maxMana,
        enemy.manaRegen,
        enemy.resistances,
        enemy.loot,
        enemy.damage
    )

    constructor(entity: Entity, damage: Damage) : this(
        entity.name,
        entity.id,
        entity.health,
        entity.maxHealth,
        entity.healthRegen,
        entity.mana,
        entity.maxMana,
        entity.manaRegen,
        entity.resistances,
        entity.loot,
        damage
    )

    private var tick = 1
    private val defCoefficient = 1.05

    override fun doDamage(target: Health) {
        target.takeDamage(damage)
    }

    override fun doDamage(target: Health, ref: DatabaseReference) {
        target.takeDamage(damage, ref)
    }

    fun attack(target: Health) {
        if (tick % 2 == 0) {
            defend()
            doDamage(target)
        } else {
            defend()
        }
        tick++
    }

    override fun defend() {
        def = !def
        if (def)
            resistances.applyDefence(defCoefficient)
        else
            resistances.removeDefence(defCoefficient)
    }

    override fun takeDamage(damage: Damage, ref: DatabaseReference) {
        super.takeDamage(damage, ref)
        ref.setValue(Json.encodeToString(serializer(), this))
    }
}